torch/csrc/jit/script/compiler.cpp

```diff
@@ -1396,17 +1396,77 @@ struct to_ir {
       const Ident& target,
       const List<Expr>& args,
       const List<Stmt>& body) {
-    // TODO: start, stop, step loop
-    if (args.size() != 1) {
+    Value *end_val = nullptr, *start_val = nullptr, *step_val = nullptr;
+    bool isSimpleRange = (args.size() == 1);
+    std::vector<Value*> argVals;
+    for (auto i: args) {
+      argVals.push_back(ensureInt(range, emitExpr(i)));
+    }
+    if (isSimpleRange) {
+      end_val = argVals[0];
+      start_val = end_val->owningGraph()->insertConstant(0);
+      step_val = end_val->owningGraph()->insertConstant(1);
+    } else if (args.size() == 2 || args.size() == 3) {
+      start_val = argVals[0];
+      end_val = argVals[1];
+      if (args.size() == 3) {
+        step_val = argVals[2];
+      } else {
+        step_val = end_val->owningGraph()->insertConstant(1);
+      }
+    } else if (args.size() == 0) {
+      throw ErrorReport(range)
+          << "range expected 1 arguments, got 0";
+    } else {
       throw ErrorReport(range)
-          << "range() expects 1 argument but got " << args.size();
+          << "range expected at most 3 arguments, got " << args.size();
     }
-    auto max_trip_count_val = ensureInt(range, emitExpr(args[0]));
     const auto& ident_name = target.name();
-    auto assigner = [ident_name, range](
+    AT_CHECK(end_val != nullptr && start_val != nullptr && step_val != nullptr, "Expected non-null pointers for range() arguments");
+    auto addOp = [end_val](Graph *g, NodeKind x, ArrayRef<Value*> inputs) {
+      return g->insertNode(g->create(x, inputs, 1))
+          ->setSourceRange(end_val->node()->sourceRange())
+          ->output()
+          ->setType(IntType::get());
+    };
+    auto assigner = [addOp, ident_name, range, start_val, step_val, isSimpleRange](
                         Value* index, std::shared_ptr<Environment> env) {
-      env->setVar(range, ident_name, index);
+      Value* derived_index;
+      if (isSimpleRange) {
+        derived_index = index;
+      } else {
+        auto g = index->owningGraph();
+        derived_index = addOp(g, aten::__derive_index, {index, start_val, step_val});
+      }
+      env->setVar(range, ident_name, derived_index);
     };
+    Value* max_trip_count_val;
+    if (isSimpleRange) {
+      max_trip_count_val = end_val;
+    } else {
+      auto g = start_val->owningGraph();
+      Value* cond_value = emitBuiltinCall(
+            range,
+            *g,
+            aten::eq,
+            c10::nullopt,
+            {step_val, g->insertConstant(0)},
+            {},
+          /*required=*/true);
+      Node* n = g->insertNode(create(prim::If, range, 0));
```

Comment:
This would make much more sense as an error in the operator. The extra interpreter overhead to handle this if statement is going to be much worse _and_ this is another instance of generating too much code using graph APIs in the compiler.

Reply:
in the vast majority of cases the step will be constant and we can compile it away